### PR TITLE
zfs_unstable: 2.2.6 → 2.3.0-rc2

### PR DESCRIPTION
**Pull request by toastal**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

[^1]

[^1]: Please consider [giving up MS GitHub](https://sfconservancy.org/GiveUpGitHub/) or offering a non-proprietary, non-US-corporate-controlled mirror for this free software project. I wish to delete this Microsoft account in the future, but I need more projects like this to support alternative methods to send patches & contribute.

### CODE REVIEW

**Comment by Titaniumtown** [comment]
@toastal if you rebase I can test it locally :)

File: pkgs/os-specific/linux/zfs/unstable.nix [hunk 1]
```diff
@@ -20,12 +20,12 @@ callPackage ./generic.nix args {
   # IMPORTANT: Always use a tagged release candidate or commits from the
   # zfs-<version>-staging branch, because this is tested by the OpenZFS
   # maintainers.
-  version = "2.2.6";
-  # rev = "";
+  version = "2.3.0-rc2";
+  rev = "zfs-2.3.0-rc2";
```
Comment (Titaniumtown) [inline pkgs/os-specific/linux/zfs/unstable.nix:24]: On line 17, `kernelCompatible = kernel: kernel.kernelOlder "6.11";` should be anything before 6.12 as 2.3.0-rc2 supports 6.11

**Review by Titaniumtown — CHANGES_REQUESTED** [review]
Seems there are more changes when it comes to building:
```shell
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/zvii3zim2cry704j515j5m9nncag19g2-source
source root is source
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
substituteStream() in derivation zfs-user-zfs-2.3.0-rc2: WARNING: '--replace' is deprecated, use --replace-{fail,warn,quiet}. (file './lib/libshare/os/linux/nfs.c')
substitute(): ERROR: file './etc/zfs/Makefile.am' does not exist
```

**Comment by Titaniumtown** [comment]
> These older functions are an off-by-one problem in my head. Good catch. However it did build build & run locally for me on the other front which is a bit concerning. I will try to look when I am free again, but I may be busy til the weekend

I rebased your patches on master locally to test, that is maybe why I got a different result.

**Comment by surfaceflinger** [comment]
I would wait with this for now https://github.com/openzfs/zfs/issues/16631

**Comment by surfaceflinger** [comment]
ZFS on LUKS is unfortunately quite common, idk why, maybe because ZFS native encryption is still too new for some people and also leaks metadata

**Comment by surfaceflinger** [comment]
also wouldn't --replace-fail be better? Personally I never used --replace-warn because I wouldn't like ryantm bot to open a PR with broken replaces for someone to then merge it blindly

**Comment by adamcstephens** [comment]
> I am not a maintainer nor did I write the code so I am not touching it. I just want it to not fail on build & be a reminder to the maintainer to fix the warn if they don’t like it.

You did touch it though. It should indeed fail, not warn.

**Comment by adamcstephens** [comment]
> It seems unreasonable to make contributors take on the burden of the maintainer in a separate file from the proposed upgrade

Part of being a contributor is taking that burden on yourself to do things the correct way, even if you're not a maintainer of the package. Sometimes that means touching files outside the simple change you wish to make. If that makes you uncomfortable, that's ok, you have the option of not making the contribution. I definitely walk away from contributions that grow past what I'm ready to commit to changing.

In this case, the file is a common include for multiple versions. As a maintainer of the stable zfs version, I disagree that `warn` is acceptable without more justification on individual calls, which should be documented in the code. That will require you to validate the affected versions and is the right way to address this automated change. It is indeed more work, but now is the time to fix it if you want to get this version updated. If you don't want to, maybe @amarshall will for you, or I will whenever the next stable release of zfs lands.

**Review by adamcstephens — CHANGES_REQUESTED** [review]
warns should be fails.

**Comment by adamcstephens** [comment]
> If amarshall chimed in I would make that change. What is the point of deprecating replace or even having --replace-warn if all replaces are supposed to be failures?

Why does my request not count? It affects a package I maintain (along with amarshall).

I didn't say they couldn't be warns. I said if they need to be warns (and some in this file even may!) that they should be correctly identified and documented. The default should be fail though, or upstream changes could be missed.

**Comment by poscat0x04** [comment]
> ZFS on LUKS is unfortunately quite common, idk why, maybe because ZFS native encryption is still too new for some people and also leaks metadata

For one zfs encryption does not support multiple key slots, which makes it hard/impossible to do stuff like using usb sticks and/or tpms to unlock your dataset, recovery keys etc.

**Pull request closed without merging**